src/gpu/nufft_synthesis.cpp

```diff
@@ -10,46 +10,63 @@
+#include "nufft_util.hpp"
 
 namespace bipp {
 namespace gpu {
 
 template <typename T>
-NufftSynthesis<T>::NufftSynthesis(std::shared_ptr<ContextInternal> ctx, T tol, std::size_t nAntenna,
-                                  std::size_t nBeam, std::size_t nIntervals, std::size_t nFilter,
-                                  const BippFilter* filterHost, std::size_t nPixel, const T* lmnX,
-                                  const T* lmnY, const T* lmnZ)
+NufftSynthesis<T>::NufftSynthesis(std::shared_ptr<ContextInternal> ctx, NufftSynthesisOptions opt,
+                                  std::size_t nAntenna, std::size_t nBeam, std::size_t nIntervals,
+                                  std::size_t nFilter, const BippFilter* filterHost,
+                                  std::size_t nPixel, const T* lmnX, const T* lmnY, const T* lmnZ)
     : ctx_(std::move(ctx)),
-      tol_(tol),
+      opt_(std::move(opt)),
       nIntervals_(nIntervals),
       nFilter_(nFilter),
       nPixel_(nPixel),
       nAntenna_(nAntenna),
       nBeam_(nBeam),
-      inputCount_(0) {
+      imgPartition_(DomainPartition::none(ctx_, nPixel)),
+      collectCount_(0) {
   auto& queue = ctx_->gpu_queue();
   filterHost_ = queue.create_host_buffer<BippFilter>(nFilter_);
   std::memcpy(filterHost_.get(), filterHost, sizeof(BippFilter) * nFilter_);
+
+  std::visit(
+      [&](auto&& arg) -> void {
+        using ArgType = std::decay_t<decltype(arg)>;
+        if constexpr (std::is_same_v<ArgType, Partition::Grid>) {
+          imgPartition_ =
+              DomainPartition::grid<T>(ctx_, arg.dimensions, nPixel_, {lmnX, lmnY, lmnZ});
+        }
+      },
+      opt_.localImagePartition.method);
+
   lmnX_ = queue.create_device_buffer<T>(nPixel_);
-  api::memcpy_async(lmnX_.get(), lmnX, sizeof(T) * nPixel_, api::flag::MemcpyDeviceToDevice,
-                    queue.stream());
   lmnY_ = queue.create_device_buffer<T>(nPixel_);
-  api::memcpy_async(lmnY_.get(), lmnY, sizeof(T) * nPixel_, api::flag::MemcpyDeviceToDevice,
-                    queue.stream());
   lmnZ_ = queue.create_device_buffer<T>(nPixel_);
-  api::memcpy_async(lmnZ_.get(), lmnZ, sizeof(T) * nPixel_, api::flag::MemcpyDeviceToDevice,
-                    queue.stream());
 
-  // use at most 33% of memory more accumulation, but not more than 200
-  // iterations. TODO: find optimum
-  std::size_t freeMem, totalMem;
-  api::mem_get_info(&freeMem, &totalMem);
-  nMaxInputCount_ =
-      (totalMem / 3) / (nIntervals_ * nFilter_ * nAntenna_ * nAntenna_ * sizeof(std::complex<T>));
-  nMaxInputCount_ = std::min<std::size_t>(std::max<std::size_t>(1, nMaxInputCount_), 200);
+  imgPartition_.apply(lmnX, lmnX_.get());
+  imgPartition_.apply(lmnY, lmnY_.get());
+  imgPartition_.apply(lmnZ, lmnZ_.get());
+
+  if (opt_.collectGroupSize && opt_.collectGroupSize.value() > 0) {
+    nMaxInputCount_ = opt_.collectGroupSize.value();
+  } else {
+    // use at most 25% of memory more accumulation, but not more than 200
```

Comment:
Could be dangerous to base the estimate on the total memory regardless of the actual usage

Reply:
The problem with using available memory is that the result changes depending on what happened beforehand, which could lead to unpredictable performance degradation. And since we use a memory pool, actual allocations are not freed, which would then impact the second imager, even though the memory pool likely has the same amount of memory available. So I think throwing an error is a better way, since allocation happen very early in the execution and then the user has the option to tune parameters manually if required.

Reply:
When I faced the issue it was on the second imager. But in that case, the problem could be solved by deleting and recreating the context so that the destructor of the memory pool frees the memory.
Otherwise we could have a function like `ctx->free_memory()` that would reset the memory pool, but that manual intervention, so not ideal either.
For me it's fine to go ahead.